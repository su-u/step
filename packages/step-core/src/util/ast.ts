import { IInterpreterRules } from '../interpriterRules/types';

export const getIndex = ({ ast, execObject: { manager }}: IInterpreterRules<any>) => {
  if (ast.children.NumberLiteral) {
    return parseInt(ast.children.NumberLiteral[0].image);
  } else if (ast.children.Identifier) {
    return manager.variable.reference(ast.children.Identifier[0].image).image;
  }
};
