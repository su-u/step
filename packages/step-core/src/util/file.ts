import * as fs from 'fs-extra';

export const writeAstToJson = (ast: any): void => {
  fs.writeFile('out.json', JSON.stringify(ast, undefined, 2), (err) => {
    if (err) {
      console.error(err);
    } else {
      console.info('write end');
    }
  });
};
