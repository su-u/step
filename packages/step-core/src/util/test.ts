import { parseInput } from '../parser';
import { Manager } from '../manager';
import { interpreter } from '../interpreter';

export const sourceParseJson = (source: string) => {
  return JSON.stringify(parseInput(source), undefined, 2);
};

export const exec = (text: string, manager: Manager) => {
  const ast = parseInput(text);
  interpreter({ ast, context: { manager, interpreter } });
  return manager;
};
