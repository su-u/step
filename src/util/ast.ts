import { IInterpreterRules } from '../interpriterRules/types';

export const getIndex = ({ ast, manager }: IInterpreterRules) => {
  if (ast.children.NumberLiteral !== undefined) {
    return parseInt(ast.children.NumberLiteral[0].image);
  } else if (ast.children.Identifier !== undefined) {
    return manager.variable.reference(ast.children.Identifier[0].image).image;
  }
};
