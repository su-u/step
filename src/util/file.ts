import * as fs from 'fs-extra';
import logger from '@/logger';

export const writeAstToJson = (ast: AstType): void => {
  fs.writeFile('out.json', JSON.stringify(ast, undefined, 2), (err) => {
    if (err) {
      logger.error(err);
    } else {
      logger.info('write end');
    }
  });
};
