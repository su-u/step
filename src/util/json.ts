export const toJson = (ast: any) => {
  return JSON.stringify(ast, undefined, 2);
};
