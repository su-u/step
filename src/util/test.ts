import { parseInput } from '@/parser';

export const sourceParse = (source: string) => {
  return JSON.stringify(parseInput(source), undefined, 2);
};
