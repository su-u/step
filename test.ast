Ruleset = syntax.pegjs
source code = example4.chb
{
  type: 'Program',
  body: [
    {
      type: 'ExpressionStatement',
      expression: {
        type: 'AssignmentExpression',
        operator: '=',
        left: { type: 'Identifier', name: 'a' },
        right: { type: 'Literal', value: 'foo', class: 'String' }
      }
    },
    {
      type: 'ExpressionStatement',
      expression: {
        type: 'AssignmentExpression',
        operator: '=',
        left: { type: 'Identifier', name: 'b' },
        right: { type: 'Literal', value: 'bar', class: 'String' }
      }
    },
    {
      type: 'ExpressionStatement',
      expression: {
        type: 'AssignmentExpression',
        operator: '=',
        left: { type: 'Identifier', name: 'c' },
        right: {
          type: 'BinaryExpression',
          operator: '+',
          left: { type: 'Identifier', name: 'a' },
          right: { type: 'Identifier', name: 'b' }
        }
      }
    },
    {
      type: 'ExpressionStatement',
      expression: {
        type: 'AssignmentExpression',
        operator: '=',
        left: { type: 'Identifier', name: 'd' },
        right: {
          type: 'BinaryExpression',
          operator: '+',
          left: { type: 'Literal', value: 1, class: 'Number' },
          right: { type: 'Literal', value: 5, class: 'Number' }
        }
      }
    },
    {
      type: 'ExpressionStatement',
      expression: {
        type: 'AssignmentExpression',
        operator: '=',
        left: { type: 'Identifier', name: 'e' },
        right: {
          type: 'BinaryExpression',
          operator: '+',
          left: { type: 'Identifier', name: 'c' },
          right: { type: 'Identifier', name: 'd' }
        }
      }
    },
    {
      type: 'ExpressionStatement',
      expression: {
        type: 'AssignmentExpression',
        operator: '=',
        left: { type: 'Identifier', name: 'f' },
        right: {
          type: 'BinaryExpression',
          operator: '>',
          left: { type: 'Literal', value: 3, class: 'Number' },
          right: { type: 'Literal', value: 4, class: 'Number' }
        }
      }
    },
    {
      type: 'ExpressionStatement',
      expression: {
        type: 'AssignmentExpression',
        operator: '=',
        left: { type: 'Identifier', name: 'g' },
        right: {
          type: 'BinaryExpression',
          operator: '<',
          left: { type: 'Literal', value: 3, class: 'Number' },
          right: { type: 'Literal', value: 4, class: 'Number' }
        }
      }
    },
    {
      type: 'ExpressionStatement',
      expression: {
        type: 'AssignmentExpression',
        operator: '=',
        left: { type: 'Identifier', name: 'h' },
        right: {
          type: 'BinaryExpression',
          operator: '<=',
          left: {
            type: 'BinaryExpression',
            operator: '+',
            left: { type: 'Literal', value: 1, class: 'Number' },
            right: {
              type: 'BinaryExpression',
              operator: '*',
              left: { type: 'Literal', value: 2, class: 'Number' },
              right: { type: 'Literal', value: 3, class: 'Number' }
            }
          },
          right: { type: 'Literal', value: 5, class: 'Number' }
        }
      }
    },
    {
      type: 'ExpressionStatement',
      expression: {
        type: 'AssignmentExpression',
        operator: '=',
        left: { type: 'Identifier', name: 'i' },
        right: {
          type: 'BinaryExpression',
          operator: '=',
          left: {
            type: 'BinaryExpression',
            operator: '/',
            left: { type: 'Literal', value: 8, class: 'Number' },
            right: { type: 'Literal', value: 4, class: 'Number' }
          },
          right: {
            type: 'BinaryExpression',
            operator: '/',
            left: { type: 'Literal', value: 18, class: 'Number' },
            right: { type: 'Literal', value: 9, class: 'Number' }
          }
        }
      }
    },
    {
      type: 'ExpressionStatement',
      expression: {
        type: 'AssignmentExpression',
        operator: '=',
        left: { type: 'Identifier', name: 'j' },
        right: {
          type: 'BinaryExpression',
          operator: '=',
          left: { type: 'Literal', value: '6', class: 'String' },
          right: { type: 'Identifier', name: 'd' }
        }
      }
    }
  ]
}
