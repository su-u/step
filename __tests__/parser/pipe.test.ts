import { sourceParseJson } from '@/util/test';

describe('パイプ', () => {
  test('1', () => {
    const source = `
"test" -> value
value |> console
    `;
    expect(sourceParseJson(source)).toMatchSnapshot();
  });

  test('2', () => {
    const source = `
"test" -> value
value |> console |> console
    `;
    expect(sourceParseJson(source)).toMatchSnapshot();
  });
});
