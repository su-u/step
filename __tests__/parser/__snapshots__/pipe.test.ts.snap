// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`パイプ 1 1`] = `
"{
  \\"name\\": \\"Program\\",
  \\"children\\": {
    \\"Assignment\\": [
      {
        \\"name\\": \\"Assignment\\",
        \\"children\\": {
          \\"ToRight\\": [
            {
              \\"name\\": \\"ToRight\\",
              \\"children\\": {
                \\"from\\": [
                  {
                    \\"name\\": \\"Pipe\\",
                    \\"children\\": {
                      \\"from\\": [
                        {
                          \\"name\\": \\"PipeFrom\\",
                          \\"children\\": {
                            \\"RelationExpression\\": [
                              {
                                \\"name\\": \\"RelationExpression\\",
                                \\"children\\": {
                                  \\"Expression\\": [
                                    {
                                      \\"name\\": \\"Expression\\",
                                      \\"children\\": {
                                        \\"Term\\": [
                                          {
                                            \\"name\\": \\"Term\\",
                                            \\"children\\": {
                                              \\"Factor\\": [
                                                {
                                                  \\"name\\": \\"Factor\\",
                                                  \\"children\\": {
                                                    \\"NumberLiteral\\": [
                                                      {
                                                        \\"image\\": \\"1\\",
                                                        \\"startOffset\\": 5,
                                                        \\"endOffset\\": 5,
                                                        \\"startLine\\": 2,
                                                        \\"endLine\\": 2,
                                                        \\"startColumn\\": 5,
                                                        \\"endColumn\\": 5,
                                                        \\"tokenTypeIdx\\": 4,
                                                        \\"tokenType\\": {
                                                          \\"name\\": \\"NumberLiteral\\",
                                                          \\"PATTERN\\": {},
                                                          \\"tokenTypeIdx\\": 4,
                                                          \\"CATEGORIES\\": [],
                                                          \\"categoryMatches\\": [],
                                                          \\"categoryMatchesMap\\": {},
                                                          \\"isParent\\": false
                                                        }
                                                      }
                                                    ]
                                                  }
                                                }
                                              ]
                                            }
                                          }
                                        ]
                                      }
                                    }
                                  ]
                                }
                              }
                            ]
                          }
                        }
                      ],
                      \\"PipeToken\\": [
                        {
                          \\"image\\": \\"|>\\",
                          \\"startOffset\\": 7,
                          \\"endOffset\\": 8,
                          \\"startLine\\": 2,
                          \\"endLine\\": 2,
                          \\"startColumn\\": 7,
                          \\"endColumn\\": 8,
                          \\"tokenTypeIdx\\": 34,
                          \\"tokenType\\": {
                            \\"name\\": \\"PipeToken\\",
                            \\"PATTERN\\": {},
                            \\"tokenTypeIdx\\": 34,
                            \\"CATEGORIES\\": [],
                            \\"categoryMatches\\": [],
                            \\"categoryMatchesMap\\": {},
                            \\"isParent\\": false
                          }
                        }
                      ],
                      \\"toIdentifier\\": [
                        {
                          \\"image\\": \\"console\\",
                          \\"startOffset\\": 10,
                          \\"endOffset\\": 16,
                          \\"startLine\\": 2,
                          \\"endLine\\": 2,
                          \\"startColumn\\": 10,
                          \\"endColumn\\": 16,
                          \\"tokenTypeIdx\\": 8,
                          \\"tokenType\\": {
                            \\"name\\": \\"Identifier\\",
                            \\"PATTERN\\": {},
                            \\"tokenTypeIdx\\": 8,
                            \\"CATEGORIES\\": [],
                            \\"categoryMatches\\": [],
                            \\"categoryMatchesMap\\": {},
                            \\"isParent\\": false
                          }
                        }
                      ]
                    }
                  }
                ]
              }
            }
          ]
        }
      }
    ]
  }
}"
`;

exports[`パイプ 2 1`] = `
"{
  \\"name\\": \\"Program\\",
  \\"children\\": {
    \\"Assignment\\": [
      {
        \\"name\\": \\"Assignment\\",
        \\"children\\": {
          \\"ToRight\\": [
            {
              \\"name\\": \\"ToRight\\",
              \\"children\\": {
                \\"from\\": [
                  {
                    \\"name\\": \\"Pipe\\",
                    \\"children\\": {
                      \\"from\\": [
                        {
                          \\"name\\": \\"PipeFrom\\",
                          \\"children\\": {
                            \\"RelationExpression\\": [
                              {
                                \\"name\\": \\"RelationExpression\\",
                                \\"children\\": {
                                  \\"Expression\\": [
                                    {
                                      \\"name\\": \\"Expression\\",
                                      \\"children\\": {
                                        \\"Term\\": [
                                          {
                                            \\"name\\": \\"Term\\",
                                            \\"children\\": {
                                              \\"Factor\\": [
                                                {
                                                  \\"name\\": \\"Factor\\",
                                                  \\"children\\": {
                                                    \\"NumberLiteral\\": [
                                                      {
                                                        \\"image\\": \\"1\\",
                                                        \\"startOffset\\": 5,
                                                        \\"endOffset\\": 5,
                                                        \\"startLine\\": 2,
                                                        \\"endLine\\": 2,
                                                        \\"startColumn\\": 5,
                                                        \\"endColumn\\": 5,
                                                        \\"tokenTypeIdx\\": 4,
                                                        \\"tokenType\\": {
                                                          \\"name\\": \\"NumberLiteral\\",
                                                          \\"PATTERN\\": {},
                                                          \\"tokenTypeIdx\\": 4,
                                                          \\"CATEGORIES\\": [],
                                                          \\"categoryMatches\\": [],
                                                          \\"categoryMatchesMap\\": {},
                                                          \\"isParent\\": false
                                                        }
                                                      }
                                                    ]
                                                  }
                                                }
                                              ]
                                            }
                                          },
                                          {
                                            \\"name\\": \\"Term\\",
                                            \\"children\\": {
                                              \\"Factor\\": [
                                                {
                                                  \\"name\\": \\"Factor\\",
                                                  \\"children\\": {
                                                    \\"NumberLiteral\\": [
                                                      {
                                                        \\"image\\": \\"1\\",
                                                        \\"startOffset\\": 9,
                                                        \\"endOffset\\": 9,
                                                        \\"startLine\\": 2,
                                                        \\"endLine\\": 2,
                                                        \\"startColumn\\": 9,
                                                        \\"endColumn\\": 9,
                                                        \\"tokenTypeIdx\\": 4,
                                                        \\"tokenType\\": {
                                                          \\"name\\": \\"NumberLiteral\\",
                                                          \\"PATTERN\\": {},
                                                          \\"tokenTypeIdx\\": 4,
                                                          \\"CATEGORIES\\": [],
                                                          \\"categoryMatches\\": [],
                                                          \\"categoryMatchesMap\\": {},
                                                          \\"isParent\\": false
                                                        }
                                                      }
                                                    ]
                                                  }
                                                }
                                              ]
                                            }
                                          }
                                        ],
                                        \\"AdditionOperator\\": [
                                          {
                                            \\"image\\": \\"+\\",
                                            \\"startOffset\\": 7,
                                            \\"endOffset\\": 7,
                                            \\"startLine\\": 2,
                                            \\"endLine\\": 2,
                                            \\"startColumn\\": 7,
                                            \\"endColumn\\": 7,
                                            \\"tokenTypeIdx\\": 18,
                                            \\"tokenType\\": {
                                              \\"name\\": \\"Plus\\",
                                              \\"PATTERN\\": {},
                                              \\"CATEGORIES\\": [
                                                {
                                                  \\"name\\": \\"AdditionOperator\\",
                                                  \\"PATTERN\\": {},
                                                  \\"tokenTypeIdx\\": 17,
                                                  \\"CATEGORIES\\": [],
                                                  \\"categoryMatches\\": [
                                                    18,
                                                    19
                                                  ],
                                                  \\"categoryMatchesMap\\": {
                                                    \\"18\\": true,
                                                    \\"19\\": true
                                                  },
                                                  \\"isParent\\": true
                                                }
                                              ],
                                              \\"tokenTypeIdx\\": 18,
                                              \\"categoryMatches\\": [],
                                              \\"categoryMatchesMap\\": {},
                                              \\"isParent\\": false
                                            }
                                          }
                                        ]
                                      }
                                    }
                                  ]
                                }
                              }
                            ]
                          }
                        }
                      ],
                      \\"PipeToken\\": [
                        {
                          \\"image\\": \\"|>\\",
                          \\"startOffset\\": 11,
                          \\"endOffset\\": 12,
                          \\"startLine\\": 2,
                          \\"endLine\\": 2,
                          \\"startColumn\\": 11,
                          \\"endColumn\\": 12,
                          \\"tokenTypeIdx\\": 34,
                          \\"tokenType\\": {
                            \\"name\\": \\"PipeToken\\",
                            \\"PATTERN\\": {},
                            \\"tokenTypeIdx\\": 34,
                            \\"CATEGORIES\\": [],
                            \\"categoryMatches\\": [],
                            \\"categoryMatchesMap\\": {},
                            \\"isParent\\": false
                          }
                        }
                      ],
                      \\"toIdentifier\\": [
                        {
                          \\"image\\": \\"console\\",
                          \\"startOffset\\": 14,
                          \\"endOffset\\": 20,
                          \\"startLine\\": 2,
                          \\"endLine\\": 2,
                          \\"startColumn\\": 14,
                          \\"endColumn\\": 20,
                          \\"tokenTypeIdx\\": 8,
                          \\"tokenType\\": {
                            \\"name\\": \\"Identifier\\",
                            \\"PATTERN\\": {},
                            \\"tokenTypeIdx\\": 8,
                            \\"CATEGORIES\\": [],
                            \\"categoryMatches\\": [],
                            \\"categoryMatchesMap\\": {},
                            \\"isParent\\": false
                          }
                        }
                      ]
                    }
                  }
                ]
              }
            }
          ]
        }
      }
    ]
  }
}"
`;

exports[`パイプ 3 1`] = `
"{
  \\"name\\": \\"Program\\",
  \\"children\\": {
    \\"Assignment\\": [
      {
        \\"name\\": \\"Assignment\\",
        \\"children\\": {
          \\"ToLeft\\": [
            {
              \\"name\\": \\"ToLeft\\",
              \\"children\\": {
                \\"to\\": [
                  {
                    \\"image\\": \\"value\\",
                    \\"startOffset\\": 5,
                    \\"endOffset\\": 9,
                    \\"startLine\\": 2,
                    \\"endLine\\": 2,
                    \\"startColumn\\": 5,
                    \\"endColumn\\": 9,
                    \\"tokenTypeIdx\\": 8,
                    \\"tokenType\\": {
                      \\"name\\": \\"Identifier\\",
                      \\"PATTERN\\": {},
                      \\"tokenTypeIdx\\": 8,
                      \\"CATEGORIES\\": [],
                      \\"categoryMatches\\": [],
                      \\"categoryMatchesMap\\": {},
                      \\"isParent\\": false
                    }
                  }
                ],
                \\"ToLeftToken\\": [
                  {
                    \\"image\\": \\"<-\\",
                    \\"startOffset\\": 11,
                    \\"endOffset\\": 12,
                    \\"startLine\\": 2,
                    \\"endLine\\": 2,
                    \\"startColumn\\": 11,
                    \\"endColumn\\": 12,
                    \\"tokenTypeIdx\\": 37,
                    \\"tokenType\\": {
                      \\"name\\": \\"ToLeftToken\\",
                      \\"PATTERN\\": {},
                      \\"tokenTypeIdx\\": 37,
                      \\"CATEGORIES\\": [],
                      \\"categoryMatches\\": [],
                      \\"categoryMatchesMap\\": {},
                      \\"isParent\\": false
                    }
                  }
                ],
                \\"from\\": [
                  {
                    \\"name\\": \\"Pipe\\",
                    \\"children\\": {
                      \\"from\\": [
                        {
                          \\"name\\": \\"PipeFrom\\",
                          \\"children\\": {
                            \\"RelationExpression\\": [
                              {
                                \\"name\\": \\"RelationExpression\\",
                                \\"children\\": {
                                  \\"Expression\\": [
                                    {
                                      \\"name\\": \\"Expression\\",
                                      \\"children\\": {
                                        \\"Term\\": [
                                          {
                                            \\"name\\": \\"Term\\",
                                            \\"children\\": {
                                              \\"Factor\\": [
                                                {
                                                  \\"name\\": \\"Factor\\",
                                                  \\"children\\": {
                                                    \\"StringLiteral\\": [
                                                      {
                                                        \\"image\\": \\"\\\\\\"test\\\\\\"\\",
                                                        \\"startOffset\\": 14,
                                                        \\"endOffset\\": 19,
                                                        \\"startLine\\": 2,
                                                        \\"endLine\\": 2,
                                                        \\"startColumn\\": 14,
                                                        \\"endColumn\\": 19,
                                                        \\"tokenTypeIdx\\": 3,
                                                        \\"tokenType\\": {
                                                          \\"name\\": \\"StringLiteral\\",
                                                          \\"PATTERN\\": {},
                                                          \\"tokenTypeIdx\\": 3,
                                                          \\"CATEGORIES\\": [],
                                                          \\"categoryMatches\\": [],
                                                          \\"categoryMatchesMap\\": {},
                                                          \\"isParent\\": false
                                                        }
                                                      }
                                                    ]
                                                  }
                                                }
                                              ]
                                            }
                                          }
                                        ]
                                      }
                                    }
                                  ]
                                }
                              }
                            ]
                          }
                        }
                      ]
                    }
                  }
                ]
              }
            }
          ]
        }
      },
      {
        \\"name\\": \\"Assignment\\",
        \\"children\\": {
          \\"ToRight\\": [
            {
              \\"name\\": \\"ToRight\\",
              \\"children\\": {
                \\"from\\": [
                  {
                    \\"name\\": \\"Pipe\\",
                    \\"children\\": {
                      \\"from\\": [
                        {
                          \\"name\\": \\"PipeFrom\\",
                          \\"children\\": {
                            \\"RelationExpression\\": [
                              {
                                \\"name\\": \\"RelationExpression\\",
                                \\"children\\": {
                                  \\"Expression\\": [
                                    {
                                      \\"name\\": \\"Expression\\",
                                      \\"children\\": {
                                        \\"Term\\": [
                                          {
                                            \\"name\\": \\"Term\\",
                                            \\"children\\": {
                                              \\"Factor\\": [
                                                {
                                                  \\"name\\": \\"Factor\\",
                                                  \\"children\\": {
                                                    \\"Identifier\\": [
                                                      {
                                                        \\"image\\": \\"value\\",
                                                        \\"startOffset\\": 25,
                                                        \\"endOffset\\": 29,
                                                        \\"startLine\\": 3,
                                                        \\"endLine\\": 3,
                                                        \\"startColumn\\": 5,
                                                        \\"endColumn\\": 9,
                                                        \\"tokenTypeIdx\\": 8,
                                                        \\"tokenType\\": {
                                                          \\"name\\": \\"Identifier\\",
                                                          \\"PATTERN\\": {},
                                                          \\"tokenTypeIdx\\": 8,
                                                          \\"CATEGORIES\\": [],
                                                          \\"categoryMatches\\": [],
                                                          \\"categoryMatchesMap\\": {},
                                                          \\"isParent\\": false
                                                        }
                                                      }
                                                    ]
                                                  }
                                                }
                                              ]
                                            }
                                          }
                                        ]
                                      }
                                    }
                                  ]
                                }
                              }
                            ]
                          }
                        }
                      ],
                      \\"PipeToken\\": [
                        {
                          \\"image\\": \\"|>\\",
                          \\"startOffset\\": 31,
                          \\"endOffset\\": 32,
                          \\"startLine\\": 3,
                          \\"endLine\\": 3,
                          \\"startColumn\\": 11,
                          \\"endColumn\\": 12,
                          \\"tokenTypeIdx\\": 34,
                          \\"tokenType\\": {
                            \\"name\\": \\"PipeToken\\",
                            \\"PATTERN\\": {},
                            \\"tokenTypeIdx\\": 34,
                            \\"CATEGORIES\\": [],
                            \\"categoryMatches\\": [],
                            \\"categoryMatchesMap\\": {},
                            \\"isParent\\": false
                          }
                        }
                      ],
                      \\"toIdentifier\\": [
                        {
                          \\"image\\": \\"console\\",
                          \\"startOffset\\": 34,
                          \\"endOffset\\": 40,
                          \\"startLine\\": 3,
                          \\"endLine\\": 3,
                          \\"startColumn\\": 14,
                          \\"endColumn\\": 20,
                          \\"tokenTypeIdx\\": 8,
                          \\"tokenType\\": {
                            \\"name\\": \\"Identifier\\",
                            \\"PATTERN\\": {},
                            \\"tokenTypeIdx\\": 8,
                            \\"CATEGORIES\\": [],
                            \\"categoryMatches\\": [],
                            \\"categoryMatchesMap\\": {},
                            \\"isParent\\": false
                          }
                        }
                      ]
                    }
                  }
                ]
              }
            }
          ]
        }
      }
    ]
  }
}"
`;

exports[`パイプ 4 1`] = `
"{
  \\"name\\": \\"Program\\",
  \\"children\\": {
    \\"Assignment\\": [
      {
        \\"name\\": \\"Assignment\\",
        \\"children\\": {
          \\"ToLeft\\": [
            {
              \\"name\\": \\"ToLeft\\",
              \\"children\\": {
                \\"to\\": [
                  {
                    \\"image\\": \\"value\\",
                    \\"startOffset\\": 5,
                    \\"endOffset\\": 9,
                    \\"startLine\\": 2,
                    \\"endLine\\": 2,
                    \\"startColumn\\": 5,
                    \\"endColumn\\": 9,
                    \\"tokenTypeIdx\\": 8,
                    \\"tokenType\\": {
                      \\"name\\": \\"Identifier\\",
                      \\"PATTERN\\": {},
                      \\"tokenTypeIdx\\": 8,
                      \\"CATEGORIES\\": [],
                      \\"categoryMatches\\": [],
                      \\"categoryMatchesMap\\": {},
                      \\"isParent\\": false
                    }
                  }
                ],
                \\"ToLeftToken\\": [
                  {
                    \\"image\\": \\"<-\\",
                    \\"startOffset\\": 11,
                    \\"endOffset\\": 12,
                    \\"startLine\\": 2,
                    \\"endLine\\": 2,
                    \\"startColumn\\": 11,
                    \\"endColumn\\": 12,
                    \\"tokenTypeIdx\\": 37,
                    \\"tokenType\\": {
                      \\"name\\": \\"ToLeftToken\\",
                      \\"PATTERN\\": {},
                      \\"tokenTypeIdx\\": 37,
                      \\"CATEGORIES\\": [],
                      \\"categoryMatches\\": [],
                      \\"categoryMatchesMap\\": {},
                      \\"isParent\\": false
                    }
                  }
                ],
                \\"from\\": [
                  {
                    \\"name\\": \\"Pipe\\",
                    \\"children\\": {
                      \\"from\\": [
                        {
                          \\"name\\": \\"PipeFrom\\",
                          \\"children\\": {
                            \\"RelationExpression\\": [
                              {
                                \\"name\\": \\"RelationExpression\\",
                                \\"children\\": {
                                  \\"Expression\\": [
                                    {
                                      \\"name\\": \\"Expression\\",
                                      \\"children\\": {
                                        \\"Term\\": [
                                          {
                                            \\"name\\": \\"Term\\",
                                            \\"children\\": {
                                              \\"Factor\\": [
                                                {
                                                  \\"name\\": \\"Factor\\",
                                                  \\"children\\": {
                                                    \\"StringLiteral\\": [
                                                      {
                                                        \\"image\\": \\"\\\\\\"test\\\\\\"\\",
                                                        \\"startOffset\\": 14,
                                                        \\"endOffset\\": 19,
                                                        \\"startLine\\": 2,
                                                        \\"endLine\\": 2,
                                                        \\"startColumn\\": 14,
                                                        \\"endColumn\\": 19,
                                                        \\"tokenTypeIdx\\": 3,
                                                        \\"tokenType\\": {
                                                          \\"name\\": \\"StringLiteral\\",
                                                          \\"PATTERN\\": {},
                                                          \\"tokenTypeIdx\\": 3,
                                                          \\"CATEGORIES\\": [],
                                                          \\"categoryMatches\\": [],
                                                          \\"categoryMatchesMap\\": {},
                                                          \\"isParent\\": false
                                                        }
                                                      }
                                                    ]
                                                  }
                                                }
                                              ]
                                            }
                                          }
                                        ]
                                      }
                                    }
                                  ]
                                }
                              }
                            ]
                          }
                        }
                      ]
                    }
                  }
                ]
              }
            }
          ]
        }
      },
      {
        \\"name\\": \\"Assignment\\",
        \\"children\\": {
          \\"ToRight\\": [
            {
              \\"name\\": \\"ToRight\\",
              \\"children\\": {
                \\"from\\": [
                  {
                    \\"name\\": \\"Pipe\\",
                    \\"children\\": {
                      \\"from\\": [
                        {
                          \\"name\\": \\"PipeFrom\\",
                          \\"children\\": {
                            \\"RelationExpression\\": [
                              {
                                \\"name\\": \\"RelationExpression\\",
                                \\"children\\": {
                                  \\"Expression\\": [
                                    {
                                      \\"name\\": \\"Expression\\",
                                      \\"children\\": {
                                        \\"Term\\": [
                                          {
                                            \\"name\\": \\"Term\\",
                                            \\"children\\": {
                                              \\"Factor\\": [
                                                {
                                                  \\"name\\": \\"Factor\\",
                                                  \\"children\\": {
                                                    \\"Identifier\\": [
                                                      {
                                                        \\"image\\": \\"value\\",
                                                        \\"startOffset\\": 25,
                                                        \\"endOffset\\": 29,
                                                        \\"startLine\\": 3,
                                                        \\"endLine\\": 3,
                                                        \\"startColumn\\": 5,
                                                        \\"endColumn\\": 9,
                                                        \\"tokenTypeIdx\\": 8,
                                                        \\"tokenType\\": {
                                                          \\"name\\": \\"Identifier\\",
                                                          \\"PATTERN\\": {},
                                                          \\"tokenTypeIdx\\": 8,
                                                          \\"CATEGORIES\\": [],
                                                          \\"categoryMatches\\": [],
                                                          \\"categoryMatchesMap\\": {},
                                                          \\"isParent\\": false
                                                        }
                                                      }
                                                    ]
                                                  }
                                                }
                                              ]
                                            }
                                          }
                                        ]
                                      }
                                    }
                                  ]
                                }
                              }
                            ]
                          }
                        }
                      ],
                      \\"PipeToken\\": [
                        {
                          \\"image\\": \\"|>\\",
                          \\"startOffset\\": 31,
                          \\"endOffset\\": 32,
                          \\"startLine\\": 3,
                          \\"endLine\\": 3,
                          \\"startColumn\\": 11,
                          \\"endColumn\\": 12,
                          \\"tokenTypeIdx\\": 34,
                          \\"tokenType\\": {
                            \\"name\\": \\"PipeToken\\",
                            \\"PATTERN\\": {},
                            \\"tokenTypeIdx\\": 34,
                            \\"CATEGORIES\\": [],
                            \\"categoryMatches\\": [],
                            \\"categoryMatchesMap\\": {},
                            \\"isParent\\": false
                          }
                        },
                        {
                          \\"image\\": \\"|>\\",
                          \\"startOffset\\": 42,
                          \\"endOffset\\": 43,
                          \\"startLine\\": 3,
                          \\"endLine\\": 3,
                          \\"startColumn\\": 22,
                          \\"endColumn\\": 23,
                          \\"tokenTypeIdx\\": 34,
                          \\"tokenType\\": {
                            \\"name\\": \\"PipeToken\\",
                            \\"PATTERN\\": {},
                            \\"tokenTypeIdx\\": 34,
                            \\"CATEGORIES\\": [],
                            \\"categoryMatches\\": [],
                            \\"categoryMatchesMap\\": {},
                            \\"isParent\\": false
                          }
                        }
                      ],
                      \\"toIdentifier\\": [
                        {
                          \\"image\\": \\"console\\",
                          \\"startOffset\\": 34,
                          \\"endOffset\\": 40,
                          \\"startLine\\": 3,
                          \\"endLine\\": 3,
                          \\"startColumn\\": 14,
                          \\"endColumn\\": 20,
                          \\"tokenTypeIdx\\": 8,
                          \\"tokenType\\": {
                            \\"name\\": \\"Identifier\\",
                            \\"PATTERN\\": {},
                            \\"tokenTypeIdx\\": 8,
                            \\"CATEGORIES\\": [],
                            \\"categoryMatches\\": [],
                            \\"categoryMatchesMap\\": {},
                            \\"isParent\\": false
                          }
                        },
                        {
                          \\"image\\": \\"console\\",
                          \\"startOffset\\": 45,
                          \\"endOffset\\": 51,
                          \\"startLine\\": 3,
                          \\"endLine\\": 3,
                          \\"startColumn\\": 25,
                          \\"endColumn\\": 31,
                          \\"tokenTypeIdx\\": 8,
                          \\"tokenType\\": {
                            \\"name\\": \\"Identifier\\",
                            \\"PATTERN\\": {},
                            \\"tokenTypeIdx\\": 8,
                            \\"CATEGORIES\\": [],
                            \\"categoryMatches\\": [],
                            \\"categoryMatchesMap\\": {},
                            \\"isParent\\": false
                          }
                        }
                      ]
                    }
                  }
                ]
              }
            }
          ]
        }
      }
    ]
  }
}"
`;
